### DISCUSSION TITLE
Set of files as together working group in OME

### DISCUSSION THREAD

**Discussion by ladmanj** [OpenModelica/OpenModelica#10468] [Q&A] [answered]
Hello friends,
This has been most probably asked several times, but I can't find the usable answer anywhere.
I am Modelica and Openmodelica newbie.
I have never used Dymola or other environments.

I have OpenModelica v1.20.0 (64-bit) for Windows (and also for Linux at a different place) installed and am mostly happy with it.
The only thing that's bothering me repeatedly is, that I must load several model files (the files I created in the past) in the right order, otherways there are holes in the signal flow,  displaying crossed rectangles and it is not easy to fix it. The model unload unloads the matched dependencies so it makes additional harm. Then I instead close the OME window and start it again.

I would like to have a project file, containing all dependencies set in it and load it in one shot.
Maybe I'm supposed to make a package somehow and then load the whole package, but I don't see how to do that.

Can you help me?

Jakub Ladman

**Comment by casella** [comment, marked as answer]
@ladmanj, if you have several related models, you should enclose them in a package, so they can be loaded all at once, rather than loading them one at a time. Is that the case?

**Comment by ladmanj** [comment]
Most probably yes, but what is the right way to do it?J.Dne 29. 3. 2023 20:15 napsal uživatel Francesco Casella ***@***.***>:
@ladmanj, if you have several related models, you should enclose them in a package, so they can be loaded all at once, rather than loading them one at a time. Is that the case?

**Comment by adeas31** [comment]
You should create a package, press ctrl+n and then choose `Package` as `Specialization` and then you can choose whether to save the contents of the package in one file or as a directory structure by checking/unchecking the option `Save contents in one file`. Then you create model and insert it in the package you created above by specifying it in `Insert in class (optional)` field.

If you choose to save contents of package in one file then you get everything in a single file and you just load it. If you don't select to save contents in one file then you will get a directory structure and you load the top level `package.mo` file to load everything.

**Comment by ladmanj** [comment]
Hello
I somehow figured it out how to make a package from my project.
The process however wasn't smooth in any case and it was full of manual work, trial and error, starting from the assumptions that the transition "must" be a standard operation, to realizing that I must build the package structure first and then to move the actual textual source code representation by copy & paste.
Because the project was starting to be quite complex at the first place, which was the initial motivation to move on to the package style, it was pretty tedious.
In any case, the work is done. For this time. However I'm really afraid that I can easily find myself in the same unfortunate situation in the future for multiple times.
It would be awesome, if the transformation process would have been automated by the OMEdit tool.

BTW: I was able only to make the self contained package and not the directory linked package (with multiple files).
Regardless of it I think that the whole environment is a huge achievement and I'm thankful for that.

**Comment by casella** [comment]
@ladmanj as you now understand well, the best way to proceed is that first you decide on a package structure, build it, and then populate it with models by creating them in their right place from the start.

When you create a package with OMEdit, there is a checkbox "Save contents in one file" which is checked by default. If you leave it like that, you end up with everything in one single package file; if you uncheck it, then the package will be saved as a directory. The process is recursive, when you define sub-packages you can have them in a single file, or again spread into directories. In this way you can decide how granular you want the structure to be. At one extreme, one file for the whole package, at the other end one file for each model, with all the possible combinations in-between.

What is still missing in OMEdit is a proper "Rename" function, that would allow to also move around models from the top global scope to some package or sub-package, or from one package to another one, without breaking the lookup of class names. This would have allowed you to move the individual models into the empty package structure effortlessly. This is planned for future versions of OMEdit.

We do have a Duplicate function that could have also be used for that, but it's still not good enough, as it makes a mere copy of the model without checking for the lookup of relative pathnames, which could end up being broken. This is also going to be improved in future versions of OMEdit.
